From: downstream
Subject: Fix incompatible pointer types for GCC 14

GCC 14 defaults -Wincompatible-pointer-types to an error.

rb_compression_options.c: grpc_compression_algorithm_name() expects
const char** but was passed char**. (upstream grpc/grpc#37741)

rb_channel.c and rb_event_thread.c: rb_thread_create() expects
VALUE (*)(void*) but the callback functions take VALUE. Also,
rb_thread_call_without_gvl() first arg expects void*(*)(void*) but the
unblocking functions return void. Wrappers are added since these
functions are also used as ubf (3rd arg) which expects void(*)(void*).

diff -up a/src/ruby/ext/grpc/rb_compression_options.c b/src/ruby/ext/grpc/rb_compression_options.c
--- a/src/ruby/ext/grpc/rb_compression_options.c
+++ b/src/ruby/ext/grpc/rb_compression_options.c
@@ -296,7 +296,7 @@ VALUE grpc_rb_compression_options_level_
  * Fails if the enum value is invalid. */
 VALUE grpc_rb_compression_options_algorithm_value_to_name_internal(
     grpc_compression_algorithm internal_value) {
-  char* algorithm_name = NULL;
+  const char* algorithm_name = NULL;
 
   if (!grpc_compression_algorithm_name(internal_value, &algorithm_name)) {
     rb_raise(rb_eArgError, "Failed to convert algorithm value to name");
diff -up a/src/ruby/ext/grpc/rb_channel.c b/src/ruby/ext/grpc/rb_channel.c
--- a/src/ruby/ext/grpc/rb_channel.c
+++ b/src/ruby/ext/grpc/rb_channel.c
@@ -726,7 +726,7 @@ static void run_poll_channels_loop_unblo
 }
 
 // Poll channel connectivity states in background thread without the GIL.
-static VALUE run_poll_channels_loop(VALUE arg) {
+static VALUE run_poll_channels_loop(void* arg) {
   (void)arg;
   gpr_log(
       GPR_DEBUG,
@@ -736,6 +736,12 @@ static VALUE run_poll_channels_loop(VALU
   return Qnil;
 }
 
+// Wrapper returning void* for use as first arg to rb_thread_call_without_gvl
+static void* run_poll_channels_loop_unblocking_func_wrapper(void* arg) {
+  run_poll_channels_loop_unblocking_func(arg);
+  return NULL;
+}
+
 static void* set_abort_channel_polling_without_gil(void* arg) {
   (void)arg;
   gpr_mu_lock(&global_connection_polling_mu);
@@ -783,7 +789,7 @@ void grpc_rb_channel_polling_thread_stop
             "GRPC_RUBY: channel polling thread stop: thread was not started");
     return;
   }
-  rb_thread_call_without_gvl(run_poll_channels_loop_unblocking_func, NULL, NULL,
+  rb_thread_call_without_gvl(run_poll_channels_loop_unblocking_func_wrapper, NULL, NULL,
                              NULL);
   rb_funcall(g_channel_polling_thread, rb_intern("join"), 0);
   // state associated with the channel polling thread is destroyed, reset so
diff -up a/src/ruby/ext/grpc/rb_event_thread.c b/src/ruby/ext/grpc/rb_event_thread.c
--- a/src/ruby/ext/grpc/rb_event_thread.c
+++ b/src/ruby/ext/grpc/rb_event_thread.c
@@ -116,7 +116,7 @@ static void grpc_rb_event_unblocking_fun
 
 /* This is the implementation of the thread that handles auth metadata plugin
  * events */
-static VALUE grpc_rb_event_thread(VALUE arg) {
+static VALUE grpc_rb_event_thread(void* arg) {
   grpc_rb_event* event;
   (void)arg;
   while (true) {
@@ -135,6 +135,12 @@ static VALUE grpc_rb_event_thread(VALUE
   return Qnil;
 }
 
+// Wrapper returning void* for use as first arg to rb_thread_call_without_gvl
+static void* grpc_rb_event_unblocking_func_wrapper(void* arg) {
+  grpc_rb_event_unblocking_func(arg);
+  return NULL;
+}
+
 void grpc_rb_event_queue_thread_start() {
   if (!g_one_time_init_done) {
     g_one_time_init_done = true;
@@ -155,7 +161,7 @@ void grpc_rb_event_queue_thread_stop() {
             "GRPC_RUBY: call credentials thread stop: thread not running");
     return;
   }
-  rb_thread_call_without_gvl(grpc_rb_event_unblocking_func, NULL, NULL, NULL);
+  rb_thread_call_without_gvl(grpc_rb_event_unblocking_func_wrapper, NULL, NULL, NULL);
   rb_funcall(g_event_thread, rb_intern("join"), 0);
   g_event_thread = Qnil;
 }
diff -up a/src/ruby/ext/grpc/extconf.rb b/src/ruby/ext/grpc/extconf.rb
--- a/src/ruby/ext/grpc/extconf.rb
+++ b/src/ruby/ext/grpc/extconf.rb
@@ -171,7 +171,8 @@ if grpc_config == 'dbg'
   $CFLAGS << ' -O0 -ggdb3'
 end
 
-$LDFLAGS << ' -Wl,-wrap,memcpy' if linux
+# Remove -Wl,-wrap,memcpy: unnecessary on glibc >= 2.17 and causes
+# undefined symbol __wrap_memcpy on EL10 (glibc 2.39).
 # Do not statically link standard libraries on TruffleRuby as this does not work when compiling to bitcode
 if linux && RUBY_ENGINE != 'truffleruby'
   $LDFLAGS << ' -static-libgcc -static-libstdc++'
